### PR TITLE
Rename reader to follower

### LINKED ISSUE #2833: Chain DB: rename reader to follower
The name "reader" is not so descriptive for an object to follow the chain with. A better name would be "follower" or "chain follower".

The code, the comments, and the technical report will have to be updated accordingly.

### PR DESCRIPTION
**Pull request by EncodePanda**
Fixes #2833

### CODE REVIEW

**Comment by coot** [comment]
The changes one the network side looks good to me; I wonder if the consensus report needs to be update as well.

**Review by mrBliss — DISMISSED** [review]
As discussed in the Meet, the alignment still needs to be updated :slightly_smiling_face:

**Comment by mrBliss** [comment]
> I wonder if the consensus report needs to be update as well.

Apparently, the current version doesn't contain any references to readers :slightly_smiling_face:. I have actually written the section on readers in the last week, it will appear shortly in https://github.com/input-output-hk/ouroboros-network/pull/2853 (and I'll rename them to followers first).

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs [hunk 1]
```diff
@@ -537,35 +537,35 @@ iteratorClose :: IteratorId -> Model blk -> Model blk
...
+  Followers
 -------------------------------------------------------------------------------}
 
-readerExists :: CPS.ReaderId -> Model blk -> Bool
-readerExists rdrId = CPS.readerExists rdrId . cps
+followerExists :: CPS.FollowerId -> Model blk -> Bool
+followerExists rdrId = CPS.followerExists rdrId . cps
 
-checkIfReaderExists :: CPS.ReaderId -> Model blk
+checkIfFollowerExists :: CPS.FollowerId -> Model blk
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs:546]: Just sharing an observation. Alignment breaks like this one are the reason we chose to not align `->` with `::` in the `ouroboros-consensus/doc/StyleGuide.md`. (Note that we also chose to keep `::` on the first line for other reasons.)

So, as your judgement call, if you feel like you're fixing up this kind of alignment for almost all of the signatures in some module, feel free to also change all those signatures in such a module to adhere to the `StyleGuide.md`. EG in this case (depending how many of the other signatures in this file you're already touching), it'd change to:

```
checkIfFollowerExists ::
     CPS.FollowerId
  -> Model blk
  -> a
  -> Either (ChainDbError blk) a
```

Somewhat separately: in the case of a rename, I would be tempted to have separate commits for the rename and for the alignment-fixup. Not crucial, but I think that would make it easier for me personally to review, for example.

**Comment by nfrisby** [comment]
I see some strays left. Here is an example from each module in which I'm getting hits.

```
ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs:    = m { cps = CPS.deleteFollower rdrId (cps m) }
ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:data Cmd blk it rdr
ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:          (\rdr -> runPeer
ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToNode.hs:          (\rdr -> runPeerWithLimits
```

My command was `git grep -Ei -e '\brdr' -e 'rdr\b' -e '\breader[^T]' -e 'reader\b' -- 'ouroboros*hs'`

File: ouroboros-network/src/Ouroboros/Network/MockChain/ProducerState.hs [hunk 2]
```diff
@@ -129,48 +129,48 @@ findFirstPoint :: HasHeader block
 findFirstPoint ps = Chain.findFirstPoint ps . producerChain
 
 
--- | Add a new reader with the given intersection point and return the new
--- 'ReaderId'.
+-- | Add a new follower with the given intersection point and return the new
+-- 'FollowerId'.
 --
-initReader :: HasHeader block
+initFollower :: HasHeader block
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/MockChain/ProducerState.hs:135]: Missing alignment of :: and ->s.

File: ouroboros-network/protocol-tests/Test/ChainProducerState.hs [hunk 3]
```diff
@@ -217,28 +216,28 @@ instance Arbitrary ChainProducerStateForkTest where
...
+    -- shrink chain and fix up followers
+    ++ [ ChainProducerStateForkTest (ChainProducerState c' (fixupFollowerPointer c' <$> rs) nr) f
        | TestBlockChain c' <- shrink (TestBlockChain c)
        ]
     where
-      fixupReaderPointer :: Chain Block -> ReaderState Block -> ReaderState Block
-      fixupReaderPointer c' r@ReaderState{readerPoint} =
-        if pointOnChain readerPoint c'
+      fixupFollowerPointer :: Chain Block -> FollowerState Block -> FollowerState Block
+      fixupFollowerPointer c' r@FollowerState{followerPoint} =
```
Comment (nfrisby) [inline ouroboros-network/protocol-tests/Test/ChainProducerState.hs:237]: `r@FollowerState`. We agree that finding all of the `r`-to-`f` translations was likely untenable, but the `r@F` are pretty easy to catch. Let's get those. (This isn't the only one, but they're easy to grep for: `git grep -e 'r@F' -- '*hs'`.)

File: ouroboros-network/protocol-tests/Test/ChainProducerState.hs [hunk 4]
```diff
@@ -151,56 +150,56 @@ prop_switchFork (ChainProducerStateForkTest cps f) =
...
-        (zip (readerStates cps) (readerStates cps'))
+        (uncurry followerInv)
+        (zip (followerStates cps) (followerStates cps'))
   where
-    readerInv :: HasHeader block
-              => ReaderState block -> ReaderState block -> Bool
-    readerInv r r'
+    followerInv :: HasHeader block
+                => FollowerState block -> FollowerState block -> Bool
+    followerInv r r'
```
Comment (nfrisby) [inline ouroboros-network/protocol-tests/Test/ChainProducerState.hs:158]: Actually, this command

```
git grep -n -w -e r -- $(git diff 7160cfa8e HEAD --name-status | grep -E -e ^M | awk '{print $2}' ) | grep -i -e follow
```

finds the lines that have both the word `r` and the word "follow". That finds a handful of pretty clear cases. (There are some false positives in types too, but I think the `r` type variable in those hits is for "reference", not for "reader".)

Comment (nfrisby) [reply ouroboros-network/protocol-tests/Test/ChainProducerState.hs:158]: And this command similar finds occurrences of `rid` abbreviating "reader id".

```
git grep -n -w -e rid -- $(git diff 7160cfa8e HEAD --name-status | grep -E -e ^M | awk '{print $2}' )
```

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 5]
```diff
@@ -625,20 +625,20 @@ type IterRef blk m r = Reference (Opaque (TestIterator m blk)) r
...
 
 -- | Execution model
 data Model blk m r = Model
-  { dbModel      :: DBModel                blk
-  , knownIters   :: KnownIters             blk m r
-  , knownReaders :: KnownReaders           blk m r
-  , modelConfig  :: Opaque (TopLevelConfig blk)
+  { dbModel        :: DBModel                blk
+  , knownIters     :: KnownIters             blk m r
+  , knownFollowers :: KnownFollowers           blk m r
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:640]: There are 2 too many spaces before `blk m r`.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 6]
```diff
@@ -736,24 +737,24 @@ lockstep model@Model {..} cmd (At resp) = Event
   where
     (mockResp, dbModel') = step model cmd
     newIters   = RE.fromList $ zip (iters resp) (iters mockResp)
-    newReaders = RE.fromList $ zip (rdrs  resp) (rdrs  mockResp)
+    newFollowers = RE.fromList $ zip (rdrs  resp) (rdrs  mockResp)
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:740]: The `=` are out of alignment now.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 7]
```diff
@@ -736,24 +737,24 @@ lockstep model@Model {..} cmd (At resp) = Event
...
     model' = case unAt cmd of
-      -- When closing the database, all open iterators and readers are closed
+      -- When closing the database, all open iterators and followers are closed
       -- too, so forget them.
       Close -> model
         { dbModel      = dbModel'
         , knownIters   = RE.empty
-        , knownReaders = RE.empty
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:746]: The `=` are out of alignment now.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 8]
```diff
@@ -736,24 +737,24 @@ lockstep model@Model {..} cmd (At resp) = Event
...
+        , knownFollowers = RE.empty
         }
       WipeVolatileDB -> model
         { dbModel      = dbModel'
         , knownIters   = RE.empty
-        , knownReaders = RE.empty
+        , knownFollowers = RE.empty
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:752]: The `=` are out of alignment now.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 9]
```diff
@@ -736,24 +737,24 @@ lockstep model@Model {..} cmd (At resp) = Event
...
         }
       _ -> model
         { dbModel      = dbModel'
         , knownIters   = knownIters `RE.union` newIters
-        , knownReaders = knownReaders `RE.union` newReaders
+        , knownFollowers = knownFollowers `RE.union` newFollowers
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:757]: The `=` are out of alignment now.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 10]
```diff
@@ -1217,11 +1218,12 @@ data Tag =
 type EventPred m = C.Predicate (Event Blk m Symbolic) Tag
 
 -- | Convenience combinator for creating classifiers for successful commands
-successful :: (    Event Blk m Symbolic
-                -> Success Blk IteratorId ReaderId
-                -> Either Tag (EventPred m)
-              )
-           -> EventPred m
+successful ::
+  (    Event Blk m Symbolic
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:1222]: The parenthesized function type should be indented by 3 more spaces.

File: ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Server.hs [hunk 11]
```diff
@@ -73,32 +73,32 @@ chainSyncBlocksServer
...
-       , HasHeader (Header blk)
-       )
-    => Tracer m (TraceChainSyncServerEvent blk)
-    -> ChainDB m blk
-    -> Reader  m blk (WithPoint blk b)
-    -> ChainSyncServer b (Point blk) (Tip blk) m ()
-chainSyncServerForReader tracer chainDB rdr =
+chainSyncServerForFollower ::
+  forall m blk b.
+    ( IOLike m
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Server.hs:94]: The forall and the parens should both line up with the `T` in `=> Tracer` below.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/API.hs [hunk 12]
```diff
@@ -606,64 +604,64 @@ instance LedgerSupportsProtocol blk
...
 
 {-------------------------------------------------------------------------------
-  Readers
+  Followers
 -------------------------------------------------------------------------------}
 
--- | Reader
+-- | Follower
 --
--- See 'newHeaderReader' for more info.
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/API.hs:614]: Nice catch.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl.hs [hunk 13]
```diff
@@ -244,7 +244,7 @@ closeDB (CDBHandle varState) = do
     -- Only when the ChainDB was open
     whenJust mbOpenEnv $ \cdb@CDB{..} -> do
 
-      Reader.closeAllReaders     cdb
+      Follower.closeAllFollowers     cdb
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl.hs:247]: The `cdb`s don't line up anymore.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Follower.hs [hunk 14]
```diff
@@ -211,29 +211,29 @@ instructionHelper ::
...
         -- these two cases.
-        ReaderInit
+        FollowerInit
           -> return $ Left (RollBackTo GenesisPoint, Nothing)
-        ReaderInImmutableDB rollState immIt
+        FollowerInImmutableDB rollState immIt
           -> return $ Left (rollState, Just immIt)
 
-        ReaderInMem   rollState
+        FollowerInMem   rollState
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Follower.hs:228]: I don't know why the space before `rollState` was there to begin with. Please remove it.

File: ouroboros-network/src/Ouroboros/Network/MockChain/ProducerState.hs [hunk 15]
```diff
@@ -17,26 +17,26 @@ import           Data.Maybe (fromMaybe)
...
   deriving (Eq, Show)
 
--- | Readers are represented here as a relation.
+-- | Followers are represented here as a relation.
 --
-type ReaderStates block = Map ReaderId (ReaderState block)
+type FollowerStates block = Map FollowerId (FollowerState block)
 
-type ReaderId     = Int
+type FollowerId     = Int
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/MockChain/ProducerState.hs:33]: I don't know why that space was there to begin with. Please remove it.

**Comment by nfrisby** [comment]
Wow! That was a lot of lines! I found a few issues. You were very thorough! Again, I really don't think Thomas and I realized how big this would be. Thanks for persevering.

It's also possible that some of my comments were off base -- I wasn't looking at the context in the module, just the diff on GitHub. So feel free to shut me down in some of these comment threads. EG Maybe I applied the Consensus `StyleGuide.md` to a module that doesn't deserve it.

FYI, I can re-review the fix ups on this first thing tomorrow -- ie by 3 or 4pm Central European.

**Comment by nfrisby** [comment]
The last step will be to squash the commits. Like we chatted about, one big commit is fine for this 👍

**Comment by EncodePanda** [comment]
> I see some strays left. Here is an example from each module in which I'm getting hits.
> 
> ```
> ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs:    = m { cps = CPS.deleteFollower rdrId (cps m) }
> ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:data Cmd blk it rdr
> ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:          (\rdr -> runPeer
> ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToNode.hs:          (\rdr -> runPeerWithLimits
> ```
> 
> My command was `git grep -Ei -e '\brdr' -e 'rdr\b' -e '\breader[^T]' -e 'reader\b' -- 'ouroboros*hs'`
> 
> So close!

Ha! I know how they got here. After recent patch to memleak some of those variables were introduced and I did not notice that while rebasing. Thanks for the good catch!

**Comment by EncodePanda** [comment]
> Ha! I know how they got here. After recent patch to memleak some of those variables were introduced and I did not notice that while rebasing. Thanks for the good catch!

but I also see there were other few that I've missed
my grep-fu not strong enough

File: ouroboros-network/protocol-tests/Test/ChainProducerState.hs (hunk 3 above)
Comment (EncodePanda) [reply ouroboros-network/protocol-tests/Test/ChainProducerState.hs]: Thanks! The recent commit fixes it. Please also see the commit message on why I've changed `r` to `fs`

File: ouroboros-network/protocol-tests/Test/ChainProducerState.hs (hunk 4 above)
Comment (EncodePanda) [reply ouroboros-network/protocol-tests/Test/ChainProducerState.hs]: 👍

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs (hunk 5 above)
Comment (EncodePanda) [reply ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs]: 👍

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs (hunk 6 above)
Comment (EncodePanda) [reply ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs]: they were never aligned here, I'm gonna skip this one

**Comment by EncodePanda** [comment]
@nfrisby I've added few commits to this PR. Please review them. I will squash everything once I get the final 👍 on this.
Thanks for the detailed and close look on this one.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 16]
```diff
@@ -625,20 +625,20 @@ type IterRef blk m r = Reference (Opaque (TestIterator m blk)) r
 -- | Mapping between iterator references and mocked iterators
 type KnownIters blk m r = RefEnv (Opaque (TestIterator m blk)) IteratorId r
 
--- | Concrete or symbolic references to a real reader
-type ReaderRef blk m r = Reference (Opaque (TestReader m blk)) r
+-- | Concrete or symbolic references to a real follower
+type FollowerRef blk m f = Reference (Opaque (TestFollower m blk)) f
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:629]: I think this `r` stood for "reference", not for "reader". EG This module is using quickcheck-state-machine, and the README in QSM's repo uses `r` in that same way as an abstraction over `IORef`. https://github.com/advancedtelematic/quickcheck-state-machine#readme

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 17]
```diff
@@ -651,22 +651,23 @@ initModel :: TopLevelConfig blk
 initModel cfg initLedger (MaxClockSkew maxClockSkew) = Model
   { dbModel      = Model.empty initLedger maxClockSkew
   , knownIters   = RE.empty
-  , knownReaders = RE.empty
+  , knownFollowers = RE.empty
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:654]: The nearby `=` symbols need realignment.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 18]
```diff
@@ -265,15 +265,15 @@ type TestConstraints blk =
...
+deriving instance (TestConstraints blk, Eq   it, Eq   flr)
+               => Eq   (Success blk it flr)
+deriving instance (TestConstraints blk, Show it, Show flr)
+               => Show (Success blk it flr)
 
 -- | Short-hand
 type TestIterator m blk = WithEq (Iterator m blk (AllComponents blk))
 -- | Short-hand
-type TestReader   m blk = WithEq (Reader   m blk (AllComponents blk))
+type TestFollower   m blk = WithEq (Follower   m blk (AllComponents blk))
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:276]: Ah, some extra spaces here.

**Comment by nfrisby** [comment]
I rebased onto master (no conflicts, it was automatic) and fixed up a couple more alignment hiccups. Some of those hiccups you and I had fixed on our call yesterday, so I'm guessing we accidentally lost them somehow. We probably jumped to the squash without committing first our staged changes?

Anyway, my fixup commit is entirely whitespace. If I show it with the whitespace diff suppressed, it's empty:

```
$ git show -w 3cefd20c8
commit 3cefd20c8585131235a325df1d2ba9d0028cdfe4 (HEAD -> EncodePanda/rename-reader-to-follower)
Author: Nicolas Frisby <nick.frisby@iohk.io>
Date:   Thu Jan 21 15:51:38 2021 -0800

    TOSQUASH last alignment fixups
$
```

I'm going to squash it now and then merge! 🙌

**Review by nfrisby — APPROVED** [review]

**Pull request merged**